code/modules/xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,256 @@
+/obj/item/xenoartifact_labeler
+	name = "artifact labeler"
+	icon = 'icons/obj/xenoarchaeology/xenoartifact_tech.dmi'
+	icon_state = "xenoartifact_labeler"
+	desc = "A tool scientists use to label their alien bombs."
+	throw_speed = 3
+	throw_range = 5
+	w_class = WEIGHT_CLASS_TINY
+
+	///Checked trait
+	var/list/activator = list()
+	///Display names
+	var/list/activator_traits = list()
+
+	var/list/minor_trait = list()
+	var/list/minor_traits = list()
+
+	var/list/major_trait = list()
+	var/list/major_traits = list()
+
+	var/list/malfunction = list()
+	var/list/malfunction_list = list()  
+
+	///trait dialogue essentially
+	var/list/info_list = list()
+
+	///Name artifacts something pretty
+	var/sticker_name
+	///passed down to sticker
+	var/list/sticker_traits = list()
+
+	///Cooldown for stickers
+	COOLDOWN_DECLARE(sticker_cooldown)
+
+/obj/item/xenoartifact_labeler/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "XenoartifactLabeler")
+		ui.open()
+
+/obj/item/xenoartifact_labeler/ui_data(mob/user)
+	var/list/data = list()
+	data["activator"] = activator
+	data["activator_traits"] = get_trait_list_desc(activator_traits, /datum/xenoartifact_trait/activator)
+
+	data["minor_trait"] = minor_trait
+	data["minor_traits"] = get_trait_list_desc(minor_traits, /datum/xenoartifact_trait/minor)
+
+	data["major_trait"] = major_trait
+	data["major_traits"] = get_trait_list_desc(major_traits, /datum/xenoartifact_trait/major)
+
+	data["malfunction"] = malfunction
+	data["malfunction_list"] = get_trait_list_desc(malfunction_list, /datum/xenoartifact_trait/malfunction)
+
+	data["info_list"] = info_list
+
+	return data
+
+/obj/item/xenoartifact_labeler/ui_act(action, params)
+	if(..())
+		return
+	//Switch function could work here but it'd be pointless for the 1-2 cases
+	if(action == "print_traits" && COOLDOWN_FINISHED(src, sticker_cooldown))
+		COOLDOWN_START(src, sticker_cooldown, 5 SECONDS)
+		create_label(sticker_name)
+		return
+	else if(!COOLDOWN_FINISHED(src, sticker_cooldown) && isliving(loc))
+		var/mob/living/user = loc
+		to_chat(user, "<span class='warning'>The labeler is still printing.</span>")
+
+	if(action == "clear_traits")
+		clear_selection()
+		return
+
+	if(action == "change_print_name" && istext(params["name"]))
+		sticker_name = sanitize_text(params["name"])
+		return
+
+	trait_toggle(action, "activator", activator_traits, activator)
+	trait_toggle(action, "minor", minor_traits, minor_trait)
+	trait_toggle(action, "major", major_traits, major_trait)
+	trait_toggle(action, "malfunction", malfunction_list, malfunction)
+
+	. = TRUE
+	update_icon()
+
+//Get a list of all the specified trait types names, actually
+/obj/item/xenoartifact_labeler/proc/get_trait_list_desc(list/traits, trait_type)
+	trait_type = typesof(trait_type)
```

Comment:
How often does it run?
Is this possible to put typesof into static list?

Reply:
Switched to referencing global statics 👍